hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/HederaLifecyclesImpl.java

```diff
@@ -96,15 +111,155 @@ public void onUpdateWeight(
+                if (configNodeIds.contains(nodeId)) {
+                    configAddressBook.updateWeight(nodeId, stakingInfo.getWeight());
+                    configNodeIds.remove(nodeId);
+                } else {
+                    final var newStakingInfo = stakingInfosMap.getForModify(nodeNum);
+                    newStakingInfo.setWeight(0);
+                    logger.warn(
+                            "Node {} is deleted from configAddressBook during upgrade and "
+                                    + "is updated with weight zero",
+                            nodeId);
+                }
+            });
+            // for any newly added nodes that doesn't exist in state, weight should be set to 0
+            // irrespective of the weight provided in config.txt
+            if (!configNodeIds.isEmpty()) {
+                configNodeIds.forEach(nodeId -> {
+                    configAddressBook.updateWeight(nodeId, 0);
+                    logger.info(
+                            "Node {} is newly added in configAddressBook during upgrade "
+                                    + "with weight 0 in configAddressBook",
+                            nodeId);
+                });
+                // update the state with new nodes and set weight to 0
+                addAdditionalNodesToState(
+                        stakingInfosMap, configNodeIds, context.getConfiguration(), configAddressBookSize);
+            }
+        }
+    }
+
+    /**
+     * Add additional nodes to state with weight 0 and update all nodes maxStake to maxStakePerNode
+     * @param stakingInfoMap The state to update
+     * @param newNodeIds The new node ids to add
+     * @param config The configuration
+     * @param configAddressBookSize The size of the address book
+     */
+    private void addAdditionalNodesToState(
+            @NonNull final MerkleMapLike<EntityNum, MerkleStakingInfo> stakingInfoMap,
+            @NonNull final Set<NodeId> newNodeIds,
+            @NonNull final Configuration config,
+            final int configAddressBookSize) {
+        // Since PlatformContext configuration is not available here,
+        // we are using the default values here. Need to see how to use config here
+        final long maxStakePerNode = 5000000000000000000L / configAddressBookSize;
+        final var numRewardHistoryStoredPeriods = 365;
+
+        final var rewardSumHistory = new long[numRewardHistoryStoredPeriods + 1];
+        Arrays.fill(rewardSumHistory, 0L);
+
+        // Add new nodes with weight 0
+        for (final var nodeId : newNodeIds) {
+            final var id = new EntityNum((int) nodeId.id());
+            final var newNodeStakingInfo = new MerkleStakingInfo();
+            newNodeStakingInfo.setKey(id);
+            newNodeStakingInfo.setMaxStake(maxStakePerNode);
+            newNodeStakingInfo.setMinStake(0L);
+            newNodeStakingInfo.setRewardSumHistory(rewardSumHistory);
```

Comment:
I wonder if we have the same potential issue as below. What happens if somebody modifies the array? Should we create a copy per node?

Reply:
yes fixed. Thank you!